src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -357,13 +357,13 @@ public ReadOnlySequence<T> Slice(SequencePosition start, long length)
 
             if (sliceRange + (ulong)length > endRange)
             {
-                ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length);
             }
 
             int currentLength = sliceStartSegment.Memory.Length - (int)sliceStartIndex;
 
-            // Position in startSegment
-            if (currentLength <= length)
+            // Position not in startSegment
+            if (currentLength < length)
```

Comment:
Should 
pos = buffer.GetPosition(n);
s1 = buffer.Slice(pos);
s2 = buffer.Slice(pos, buffer.End);
s3 = Slice(pos, buffer.Length - n);
return absolutely the same result?

If yes than we should return condition as it was but add checking on the last segment before it.

Reply:
I sagest change `if (currentLength < length)` to `if (currentLength < length || currentLength = length and sliceStartObject  != endObject`
than `s3 = Slice(pos, buffer.Length - n);` will return absolutely the same result as `s1 = buffer.Slice(pos);`

Reply:
I will defer to @pakrym for what should be the expected behavior here.

Reply:
What do you mean by "absolutely". They should return an observably identical result.

Reply:
> What do you mean by "absolutely". They should return an observably identical result.

Sequences with the same numbers of segments.

Reply:
> Sequences with the same numbers of segments.

This test passes as expected:
```C#
[Fact]
public static void SliceAllOperationsAreEqual()
{
    var segment1 = new BufferSegment<byte>(new byte[10]);
    BufferSegment<byte> segment2 = segment1.Append(new byte[10]);

    var buffer = new ReadOnlySequence<byte>(segment1, 0, segment2, 10);

    for (int n = 0; n < 10; n++)
    {
        SequencePosition pos = buffer.GetPosition(n);
        ReadOnlySequence<byte> s1 = buffer.Slice(pos);
        ReadOnlySequence<byte> s2 = buffer.Slice(pos, buffer.End);
        ReadOnlySequence<byte> s3 = buffer.Slice(pos, buffer.Length - n);

        Assert.Equal(segment1, s1.Start.GetObject());
        Assert.Equal(segment2, s1.End.GetObject());

        Assert.Equal(n, s1.Start.GetInteger());
        Assert.Equal(10, s1.End.GetInteger());

        Assert.Equal(segment1, s2.Start.GetObject());
        Assert.Equal(segment2, s2.End.GetObject());

        Assert.Equal(n, s2.Start.GetInteger());
        Assert.Equal(10, s2.End.GetInteger());

        Assert.Equal(segment1, s3.Start.GetObject());
        Assert.Equal(segment2, s3.End.GetObject());

        Assert.Equal(n, s3.Start.GetInteger());
        Assert.Equal(10, s3.End.GetInteger());
    }

    for (int n = 10; n <= 20; n++)
    {
        SequencePosition pos = buffer.GetPosition(n);
        ReadOnlySequence<byte> s1 = buffer.Slice(pos);
        ReadOnlySequence<byte> s2 = buffer.Slice(pos, buffer.End);
        ReadOnlySequence<byte> s3 = buffer.Slice(pos, buffer.Length - n);

        Assert.Equal(segment2, s1.Start.GetObject());
        Assert.Equal(segment2, s1.End.GetObject());

        Assert.Equal(n - 10, s1.Start.GetInteger());
        Assert.Equal(10, s1.End.GetInteger());

        Assert.Equal(segment2, s2.Start.GetObject());
        Assert.Equal(segment2, s2.End.GetObject());

        Assert.Equal(n - 10, s2.Start.GetInteger());
        Assert.Equal(10, s2.End.GetInteger());

        Assert.Equal(segment2, s3.Start.GetObject());
        Assert.Equal(segment2, s3.End.GetObject());

        Assert.Equal(n - 10, s3.Start.GetInteger());
        Assert.Equal(10, s3.End.GetInteger());
    }
}
```

Reply:
> This test passes as expected:

I created test that all operations are equal and fixed where was not https://github.com/dotnet/corefx/pull/29231